hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerSelectionCriteria.java

```diff
@@ -92,15 +96,31 @@ private boolean isContainerReplicatingOrDeleting(ContainerID containerID) {
    */
   public NavigableSet<ContainerID> getCandidateContainers(
       DatanodeDetails node, long sizeMovedAlready) {
-    NavigableSet<ContainerID> containerIDSet =
-        new TreeSet<>(orderContainersByUsedBytes().reversed());
+    // Initialize containerSet for node
+    if (!setMap.containsKey(node)) {
```

Comment:
ContaienrBalancerSelectionCriteria is applicable within one iteration
1. all filtering can be done before setting to setMap
2. Next filtering can be basic including selectedContainers and shouldBeExcluded

Suggested to  use getCandidateContainers() to update setMap wrapping in another method.